AppSyncRealTimeClient/ConnectivityMonitor/ConnectivityMonitor.swift

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright 2018-2021 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Network
+
+typealias ConnectivityUpdates = (ConnectivityPath) -> Void
+
+protocol AnyConnectivityMonitor {
+    func start(connectivityUpdatesQueue: DispatchQueue, onConnectivityUpdates: @escaping ConnectivityUpdates)
+    func cancel()
+}
+
+class ConnectivityMonitor {
+
+    private let connectivityUpdatesQueue = DispatchQueue(
+        label: "com.amazonaws.ConnectivityMonitor.connectivityUpdatesQueue",
+        qos: .background
+    )
+    private var monitor: AnyConnectivityMonitor?
+
+    init(monitor: AnyConnectivityMonitor? = nil) {
+        self.monitor = monitor
+    }
+
+    func start(onUpdates: @escaping ConnectivityUpdates) {
+        if let monitor = monitor {
+            monitor.start(
+                connectivityUpdatesQueue: connectivityUpdatesQueue,
+                onConnectivityUpdates: onUpdates
+            )
+        } else if #available(iOS 12.0, *) {
+            let monitor = NetworkMonitor()
+            self.monitor = monitor
+            monitor.start(
+                connectivityUpdatesQueue: connectivityUpdatesQueue,
+                onConnectivityUpdates: onUpdates
+            )
+        }
+    }
+
+    func cancel() {
+        guard let monitor = monitor else {
+            return
+        }
+        monitor.cancel()
```

Comment:
I believe you want to set monitor to nil here so it is not unwrapped and used again on start.

Reply:
actually i don't think we should be `nil`-ing out the monitor, the behavior of `NWPathMonitor` is that a new one needs to be created each time it's started, that's why `NetworkMonitor`'s cancel() will `nil` out its internal reference to NWPathMonitor. This class `ConnectivityMonitor`'s purpose is to allow testability and support for iOS 12 and up to use NetworkMonitor. For testability, `ConnectivityMonitor`'s cancel() shouldn't `nil` out the mock network monitor, otherwise a `start` will cause the real NetworkMonitor to be used. At runtime, `monitor.cancel()` at this line will `nil` out the NWPathMonitor (responsibility of NetworkMonitor to manage) so it the existing NetworkMonitor will be reused at `start()` and within NetworkMonitor's start, it will create a new instance of `NWPathMonitor`